src/components/toolbar.tsx

```diff
@@ -72,10 +72,14 @@ export class ToolbarComponent extends BaseComponent<IProps, IState> {
     const renderToolButtons = (toolbarModel: IToolbarModel) => {
       const { ui: { selectedTileIds } } = this.stores;
       return toolbarModel.map(toolButton => {
+        const { document } = this.props;
+        const diagramTileCount = document.content?.getTilesOfType("Diagram").length || 0;
+        const isDisabled = (toolButton.id === "Diagram" && diagramTileCount > 0) ||
```

Comment:
From an architectural standpoint, it doesn't really make sense for the generic `ToolbarComponent` to have specific knowledge of individual components like this. Down this road lies an ever-growing list of component-specific restrictions that are special-cased in common code. The better approach here would be to have component-specific constraints configurable at the unit (or problem?) level so that they can be specified in content rather than in code. That said, I don't know what the priorities or time constraints are at this point, so maybe it makes sense to add a `TODO`/`FIXME` comment and a PT story now and put off the proper solution to later. If one were to pursue it, I think perhaps it might make sense to be in `settings` in the `ProblemConfiguration`.

Reply:
@kswenson Thanks. Are you saying this should be an option authors can turn on/off? I may be misunderstanding, but I think the story says to not allow more than one diagram tile no matter what. That's supposed to prevent problems multiple diagram tiles can introduce that we want to avoid dealing with altogether. So making it an option may not be right in this case.

Although, I'm not familiar with authoring in CLUE at all, so maybe I'm misunderstanding something.

Even if we definitely want to always prevent multiple diagram tiles, though, would it still make sense to move the check elsewhere? I think I could spend some more time on that. Maybe something to discuss at Tuesday's planning meeting.

Reply:
I think you're right that the intention is that the diagram tile is always limited to one. That said, it could make sense to move the logic elsewhere. I wouldn't be surprised if other tiles should be limited to one (dataflow comes to mind), so turning this on for any tile with a simple switch would be nice. I'm not sure if that's necessary for this PR, but it's something to consider.

Reply:
As @tealefristoe suggested, this isn't really a Diagram-tile-specific feature. Any tile could theoretically be specified as being limited to a single instance. Part of building an extensible framework is building features so that they support any client or tile rather than hardcoding the requirements of individual tiles into the framework. The rationale for making this a configuration option is so that it can be implemented without building Diagram-specific functionality into common code and so that future tiles can take advantage of the capability without making further modifications to common code.

Reply:
@kswenson @tealefristoe Thank you, both. That makes sense. I'll see what I can do.

Reply:
@kswenson @tealefristoe Am I on the right track with the latest commit? It works as intended, but seems at least a little inelegant. Maybe I can refactor some of that away, but I want to check in before bothering.

It seemed like we may as well make it possible for an author to define a set of tile types that should be limited to one tile. And then it seemed like we may as well let the author specify an exact number of instances to allow per type. Does that seem reasonable? Or is it overkill?

More importantly, it seems like a problem's settings keys can be any arbitrary value I add in an authoring file. Is that true? Or should new keys be registered somewhere that I'm missing? It doesn't feel right to use the arbitrary key name `limitedTileTypes` in the `ToolbarComponent` code since it's not defined as a specific option anywhere. (I know it may still make sense to move the check out of `ToolbarComponent`, but I think the issue will remain wherever the check is made.)

If I'm way off track with these changes, maybe we could have CLUE Story Hour on Wednesday?

Reply:
You're on the right track -- more detailed comments in the upcoming line-by-line review.